### PR TITLE
Add autocomplete

### PR DESCRIPTION
**Pull request by RasPhilCo**

### CODE REVIEW

File: src/commands/autocomplete.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    instructions: flags.boolean({description: 'display install instructions for the shell'})
+  }
+
+  async run () {
+    if (this.config.windows) {
+      this.out.warn('Autocomplete is not currently supported in Windows')
+      return
+    }
+
+    const autocompletePath = path.join(this.config.dataDir, 'client', 'autocomplete')
```
Comment (RasPhilCo) [inline src/commands/autocomplete.js]: @dickeyxxx I think I going to error here on brand new v6 installs b/c of the cli vs client dir thing we talked about earlier

File: src/commands/autocomplete.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  static hidden = true
+  static flags = {
+    commands: flags.boolean({hidden: true}),
+    script: flags.boolean({hidden: true}),
+    shell: flags.string({description: 'shell to use', char: 's'})
+  }
+
+  async run () {
+    if (this.config.windows) {
+      this.out.warn('Autocomplete is not currently supported in Windows')
```
Comment (jdx) [inline src/commands/autocomplete.js]: this is an error

File: src/commands/update.js [hunk 3]
```diff
@@ -55,4 +60,29 @@ export default class Update extends Command {
...
+    try {
+      // TODO: move from cli to client dir if not already present
+      // if (!fs.pathExistsSync(path.join(this.config.dataDir, 'client', 'autocomplete', 'bash', 'heroku'))) {
+      //   const cli = path.join(this.config.dataDir, 'cli', 'autocomplete')
+      //   const client = path.join(this.config.dataDir, 'client', 'autocomplete')
+      //   fs.copySync(cli, client)
+      // }
+      const plugins = await new Plugins(this.out).list()
+      const cmds = plugins.map(p => p.commands.map(c => {
+        if (c.hidden) return
```
Comment (jdx) [inline src/commands/update.js]: should filter this out instead of leaving nulls

File: src/commands/update.js [hunk 4]
```diff
@@ -55,4 +60,29 @@ export default class Update extends Command {
...
+        let plublicFlags = Object.keys(c.flags).filter(flag => !c.flags[flag].hidden).map(flag => `--${flag}`).join(' ')
```
Comment (jdx) [inline src/commands/update.js]: typo

File: src/commands/autocomplete.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      this.out.error('Error: Missing required argument shell')
+      return
+    }
+
+    switch (shell) {
+      case 'bash':
+        this.out.log('Symlink the autocomplete function via:')
+        this.out.log()
+        let fnFile = path.join(clientPath, 'node_modules', 'cli-engine', 'autocomplete', 'bash', 'heroku')
+        this.out.log(CustomColors.cmd(`$ ln -s ${fnFile} /usr/local/etc/bash_completion.d/heroku`))
```
Comment (jdx) [inline src/commands/autocomplete.js]: I'd love it if we had a way to verify bash is actually picking this up, but I'm not sure we can

File: src/commands/autocomplete.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+autoload -Uz compinit;
+compinit;`)
+          return
+        }
+        this.out.log('Add the autocomplete function to your fpath via:')
+        this.out.log()
+        this.out.log(CustomColors.cmd(`$ echo $(heroku autocomplete --shell zsh --script) >> ~/.zshrc`))
+        break
+      default:
+        this.out.log(`Currently ${shell} not a supported shell for autocomplete`)
```
Comment (jdx) [inline src/commands/autocomplete.js]: this is an error right?

File: src/commands/autocomplete.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      this.out.warn('Autocomplete is not currently supported in Windows')
+      return
+    }
+
+    const clientPath = path.join(this.config.dataDir, 'client')
+    if (this.flags.commands) {
+      this.out.log(path.join(clientPath, 'autocomplete', 'commands'))
+      return
+    }
+    let shell = this.flags.shell
```
Comment (jdx) [inline src/commands/autocomplete.js]: could do `const shell = this.flags.shell || this.config.shell`

File: src/commands/update.js [hunk 8]
```diff
@@ -4,6 +4,10 @@ import Command, {flags} from 'cli-engine-command'
 import Updater from '../updater'
 import PluginsUpdate from './plugins/update'
 import Analytics from '../analytics'
+import Plugins from '../plugins'
+import fs from 'fs-extra'
+import path from 'path'
+import flatten from 'lodash.flatten'
```
Comment (jdx) [inline src/commands/update.js]: could lazy-load this one

File: src/commands/update.js [hunk 9]
```diff
@@ -55,4 +60,29 @@ export default class Update extends Command {
...
+      const plugins = await new Plugins(this.out).list()
+      const cmds = plugins.map(p => p.commands.map(c => {
+        if (c.hidden) return
+        let plublicFlags = Object.keys(c.flags).filter(flag => !c.flags[flag].hidden).map(flag => `--${flag}`).join(' ')
+        let flags = plublicFlags.length ? ` ${plublicFlags}` : ''
+        let namespace = p.namespace ? `${p.namespace}:` : ''
+        return `${namespace}${c.id}${flags}`
+      }))
+      const commands = flatten(cmds).filter(c => !!c).join('\n')
+      fs.mkdirpSync(path.join(this.config.dataDir, 'client', 'autocomplete'))
```
Comment (jdx) [inline src/commands/update.js]: prefer `await fs.mkdirp()` and `await fs.writeFile()`

good to get in the habit of using async when we can

**Review by jdx — CHANGES_REQUESTED** [review]

File: src/commands/autocomplete.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  static topic = 'autocomplete'
+  static description = 'autocomplete installation instructions'
+  static hidden = true
+  static flags = {
+    commands: flags.boolean({hidden: true}),
+    script: flags.boolean({hidden: true}),
+    shell: flags.string({description: 'shell to use', char: 's'})
+  }
+
+  async run () {
```
Comment (jdx) [inline src/commands/autocomplete.js:17]: can you write a test to at least execute this code? Seeing a lot of red from codecov right now

File: src/commands/autocomplete.test.js [hunk 11]
```diff
@@ -1,8 +1,11 @@
 // @flow
 
 import Autocomplete from './autocomplete'
+import os from 'os'
 
 test('outputs commands file path', async () => {
+  // autocomplete will throw error on windows
+  if (os.platform() === 'windows') return
```
Comment (jdx) [inline src/commands/autocomplete.test.js]: hmm maybe we could do this better...

```js
let skipWindows = os.platform() === 'windows' ? xtest : test
skipWindows('outputs commands file path', async () => {
```

File: src/commands/autocomplete.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    if (this.config.windows) {
+      this.out.error('Autocomplete is not currently supported in Windows')
+      return
+    }
+
+    const autocompletePath = path.join(this.config.dataDir, 'client', 'node_modules', 'cli-engine', 'autocomplete')
```
Comment (dickeyxxx) [inline src/commands/autocomplete.js]:
```js
const autocompletePath = path.join(__dirname, '..', '..', 'autocomplete')
```

File: src/commands/update.js [hunk 13]
```diff
@@ -55,4 +59,28 @@ export default class Update extends Command {
...
+      // }
+      const plugins = await new Plugins(this.out).list()
+      const cmds = plugins.map(p => p.commands.filter(c => !c.hidden).map(c => {
+        let publicFlags = Object.keys(c.flags).filter(flag => !c.flags[flag].hidden).map(flag => `--${flag}`).join(' ')
+        let flags = publicFlags.length ? ` ${publicFlags}` : ''
+        let namespace = p.namespace ? `${p.namespace}:` : ''
+        return `${namespace}${c.id}${flags}`
+      }))
+      const commands = flatten(cmds).join('\n')
+      fs.writeFileSync(path.join(this.config.dataDir, 'client', 'node_modules', 'cli-engine', 'autocomplete', 'commands'), commands)
```
Comment (dickeyxxx) [inline src/commands/update.js]: `await fs.outputFile()`

**Review by jdx — APPROVED** [review]

**Pull request merged**